asyncua/ua/uatypes.py

```diff
@@ -329,7 +329,7 @@ def is_null(self):
     def has_null_identifier(self):
         if not self.Identifier:
             return True
-        if self.NodeIdType == NodeIdType.Guid and re.match(b'\00+', self.Identifier.bytes):
+        if self.NodeIdType == NodeIdType.Guid and self.Identifier.bytes == b'\00'*16:
```

Comment:
I know next to nothing about guid. Are you sure they are always 16 bytes?Isn't there a way to check generically that ALL bytes are 0? just wondering..

Reply:
all([i==0 for i in bytes])?

Reply:
https://reference.opcfoundation.org/v104/Core/docs/Part6/5.1.3/ Yeah, GUID are always 16 bytes.
The uuid.UUID type that is used also reinforces this. https://docs.python.org/3/library/uuid.html

Since we're using the `UUID.bytes` method, how about https://docs.python.org/3/library/uuid.html#uuid.UUID.int instead?
```python
if type(self.Identifier) == uuid.UUID and self.Identifier.int == 0:
```

Reply:
```suggestion
        if type(self.Identifier) == uuid.UUID and self.Identifier.int == 0:
```

Reply:
do we need to test for type?
the identifier MUST be an uuid.UUID, isn't it? 
In your proposition over the answer might be wrong if the type is not waht expected... Maybe we should test for type at the line over and raise omUaError if it is not the correct type
```
if not insinstance(self.identifier, uuid.UIID):
    raise RuntimeError(XXX) or UaError not sure what is bes 
if self.NodeIdType == NodeIdType.Guid and type(self.Identifier) == uuid.UUID and self.Identifier.int == 0:
```

Reply:
also in python isinstance is better than type(xx) == 11

Reply:
I don't know if you see this thread in a different context than the pull request I submitted, but I outlined the issue that requires the type check here: https://github.com/FreeOpcUa/opcua-asyncio/pull/291#issuecomment-696204533
The gist of it is that the identifier is not necessarily a UUID when the `NodeIdType` is `Guid`.
Whether that is supposed to be an allowed situation is beyond my current knowledge of this library. :D
There is a possibility that the `NodeIdType` parameter isn't even needed in the first place if the parameter and identifier type need to always be synchronous. At least I think the parameter probably shouldn't be settable arbitrarily, i. e. the checks you proposed should be in `__init__` to see if the parameter matches the identifier type, not in this method.

```suggestion
        if self.NodeIdType == NodeIdType.Guid and isinstance(self.Identifier, uuid.UUID) and self.Identifier.int == 0:
```

Reply:
OK I checked the code. Identifier must be None or UUID so you can just remove the type test

Reply:
then I can merge

Reply:
I mean I have an explicit case where that is not true. The test_unit.py test module actually uses this exact setup:
https://github.com/FreeOpcUa/opcua-asyncio/blob/9f3dd531397db854f2e1d2a59ff04f4231cab418/tests/test_unit.py#L408
So it is very possible for the identifier type and parameter to not agree, and if the resulting node ID it's used in a `create_node` the server dies - as outlined in the comment I linked.

In `__init__`, a type guess is being done only if the NodeIdType parameter isn't explicitly passed, everything else is just accepted blindly. You can pass any combination of identifier and NodeIdType, as long as it's not `None`.

That's why I think either check for disagreement at `__init__` (in which case it might show the NodeIdType param to be useless since identifier types are already unique) and get rid of the type check here or allow discrepancies but then definitely do a type check.

Reply:
Having a type check here and ignoring the case when it is not the expected type is definitely wrong.

And yes we can add type check at init but then we should check for all possible types not only the uuid case